pbm/storage/gcs/gcs.go

```diff
@@ -0,0 +1,219 @@
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+
+	"github.com/percona/percona-backup-mongodb/pbm/errors"
+	"github.com/percona/percona-backup-mongodb/pbm/log"
+	"github.com/percona/percona-backup-mongodb/pbm/storage"
+)
+
+type Config struct {
+	Bucket      string      `bson:"bucket" json:"bucket" yaml:"bucket"`
+	Prefix      string      `bson:"prefix" json:"prefix" yaml:"prefix"`
+	Credentials Credentials `bson:"credentials" json:"credentials" yaml:"credentials"`
+
+	// The maximum number of bytes that the Writer will attempt to send in a single request.
+	// https://pkg.go.dev/cloud.google.com/go/storage#Writer
+	ChunkSize *int `bson:"chunkSize,omitempty" json:"chunkSize,omitempty" yaml:"chunkSize,omitempty"`
+}
+
+func (cfg *Config) Clone() *Config {
+	if cfg == nil {
+		return nil
+	}
+
+	rv := *cfg
+	return &rv
+}
+
+type Credentials struct {
+	ProjectID  string `bson:"projectId" json:"projectId,omitempty" yaml:"projectId,omitempty"`
+	PrivateKey string `bson:"privateKey" json:"privateKey,omitempty" yaml:"privateKey,omitempty"`
+}
+
+type GCS struct {
+	opts         *Config
+	bucketHandle *gcs.BucketHandle
+	log          log.LogEvent
+}
+
+func New(opts *Config, node string, l log.LogEvent) (*GCS, error) {
+	ctx := context.Background()
+	var client *gcs.Client
+	var err error
+
+	if opts.Credentials.ProjectID != "" && opts.Credentials.PrivateKey != "" {
+		credStr := fmt.Sprintf(`{
+			"type": "service_account",
+          	"project_id": "%s",
+			"private_key": "%s",
+			"client_email": "service@%s.iam.gserviceaccount.com",
+			"auth_uri": "https://accounts.google.com/o/oauth2/auth",
+			"token_uri": "https://oauth2.googleapis.com/token",
+			"auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
+			"client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/%s.iam.gserviceaccount.com",
+			"universe_domain": "googleapis.com"
+		}`,
+			opts.Credentials.ProjectID,
+			opts.Credentials.PrivateKey,
+			opts.Credentials.ProjectID,
+			opts.Credentials.ProjectID,
+		)
+
+		credentials := option.WithCredentialsJSON([]byte(credStr))
+
+		client, err = gcs.NewClient(ctx, credentials)
+	} else {
+		client, err = gcs.NewClient(ctx)
+	}
+
+	if err != nil {
+		return nil, errors.Wrap(err, "create GCS client")
+	}
+
+	return &GCS{
+		opts:         opts,
+		bucketHandle: client.Bucket(opts.Bucket),
+		log:          l,
+	}, nil
+}
+
+func (*GCS) Type() storage.Type {
+	return storage.GCS
+}
+
+func (g *GCS) Save(name string, data io.Reader, size int64) error {
+	ctx := context.Background()
+
+	w := g.bucketHandle.Object(path.Join(g.opts.Prefix, name)).NewWriter(ctx)
+
+	if g.opts.ChunkSize != nil {
+		w.ChunkSize = *g.opts.ChunkSize
+	}
+
+	if _, err := io.Copy(w, data); err != nil {
+		return errors.Wrap(err, "save data")
+	}
+
+	if err := w.Close(); err != nil {
+		return errors.Wrap(err, "writer close")
+	}
+
+	return nil
+}
+
+func (g *GCS) SourceReader(name string) (io.ReadCloser, error) {
```

Comment:
For physical restore and S3 we have integrated this logic with `Download` (concurrently downloading file from storage). Previously that logic was applied to GCS, so we do not have it anymore. We do not need to that within this PR imo, but we need to have story for that.

Reply:
https://perconadev.atlassian.net/browse/PBM-1506